From 5efb39d3e27ef4d8ee95bd6ed1931b56b2eb7b5a Mon Sep 17 00:00:00 2001
From: Ronan Abhamon <ronan.abhamon@vates.fr>
Date: Fri, 7 May 2021 15:30:02 +0200
Subject: [PATCH] reset TD_VBD_LOG_DROPPED flag when VBD watchdog is cleared

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>
---
 drivers/tapdisk-vbd.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/tapdisk-vbd.c b/drivers/tapdisk-vbd.c
index 7e09ca7..409895d 100644
--- a/drivers/tapdisk-vbd.c
+++ b/drivers/tapdisk-vbd.c
@@ -1106,6 +1106,7 @@ void watchdog_cleared(td_vbd_t *vbd)
 		DBG(TLOG_WARN, "%s: watchdog timeout: requests were blocked\n", vbd->name);
 		/* Ideally want a direct way to flush the log */
 		tlog_precious(1);
+		td_flag_clear(vbd->state, TD_VBD_LOG_DROPPED);
 	}
 	vbd->watchdog_warned = false;
 }
-- 
2.30.1

