From 9a16f5622a9443a6cd566803ecc41963c12d5285 Mon Sep 17 00:00:00 2001
From: Anthoine Bourgeois <anthoine.bourgeois@vates.tech>
Date: Wed, 15 Jan 2025 16:33:02 +0100
Subject: [PATCH] tapdisk: use defined abstraction

Signed-off-by: Anthoine Bourgeois <anthoine.bourgeois@vates.tech>
---
 drivers/tapdisk-control.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/tapdisk-control.c b/drivers/tapdisk-control.c
index c8b9d826..506b203d 100644
--- a/drivers/tapdisk-control.c
+++ b/drivers/tapdisk-control.c
@@ -898,7 +898,7 @@ tapdisk_control_close_image(struct tapdisk_ctl_conn *conn,
 		tapdisk_nbdserver_pause(vbd->nbdserver_new, true);
 
     err = 0;
-    list_for_each_entry_safe(blkif, _blkif, &vbd->rings, entry) {
+    tapdisk_vbd_for_each_blkif(vbd, blkif, _blkif) {
 
         DPRINTF("implicitly disconnecting ring %p domid=%d, devid=%d\n",
                 blkif, blkif->domid, blkif->devid);
