From c68e06ed8235a5fd65a8fc813f899963d9fd9455 Mon Sep 17 00:00:00 2001
From: Anthoine Bourgeois <anthoine.bourgeois@vates.tech>
Date: Fri, 12 Jun 2026 14:54:33 +0200
Subject: [PATCH] tapdisk: fix header guard warnings (>=gcc-15)

Signed-off-by: Anthoine Bourgeois <anthoine.bourgeois@vates.tech>
---
 drivers/libaio-backend.h | 4 ++--
 include/util.h           | 2 +-
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/libaio-backend.h b/drivers/libaio-backend.h
index 7876c9c6..4079f70e 100644
--- a/drivers/libaio-backend.h
+++ b/drivers/libaio-backend.h
@@ -16,7 +16,7 @@
  */
 
 #ifndef LIBAIO_BACKEND_H
-#define LIBAIO_BACLEND_H
+#define LIBAIO_BACKEND_H
 
 #include <libaio.h>
 
@@ -30,4 +30,4 @@ enum {
 
 struct backend* get_libaio_backend();
 
-#endif /* LIBAIO_BACLEND_H */
+#endif /* LIBAIO_BACKEND_H */
diff --git a/include/util.h b/include/util.h
index 4834c342..f96b5f2a 100644
--- a/include/util.h
+++ b/include/util.h
@@ -16,7 +16,7 @@
  */
 
 #ifndef __TAPDISK_UTIL_H__
-#define __TAPDISK_UTIL_H_
+#define __TAPDISK_UTIL_H__
 
 #include <stddef.h>
 #include <string.h>
