From 1f886f67bb95c5c4345d9ac646690acc96569802 Mon Sep 17 00:00:00 2001
From: Anthoine Bourgeois <anthoine.bourgeois@vates.tech>
Date: Mon, 10 Feb 2025 18:54:25 +0100
Subject: [PATCH 22/24] qcow2: support commit command

Signed-off-by: Anthoine Bourgeois <anthoine.bourgeois@vates.tech>
---
 drivers/block-qcow2.c | 59 ++++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 58 insertions(+), 1 deletion(-)

diff --git a/drivers/block-qcow2.c b/drivers/block-qcow2.c
index 9ec20f5f..45902fe1 100644
--- a/drivers/block-qcow2.c
+++ b/drivers/block-qcow2.c
@@ -59,6 +59,8 @@
 #include "qapi/error.h"
 #include "sysemu/block-backend.h"
 #include "qapi/qmp/qdict.h"
+#include "qapi/qapi-commands-block-core.h"
+
 #include "tapdisk.h"
 #include "tapdisk-driver.h"
 #include "tapdisk-interface.h"
@@ -66,7 +68,7 @@
 #include "tapdisk-storage.h"
 //#include "block-crypto.h"
 
-#define DEBUGGING   0
+#define DEBUGGING   2
 
 #define __TRACE(s)							\
 	do {								\
@@ -93,6 +95,7 @@
 
 #define QCOW2_OP_READ             1
 #define QCOW2_OP_WRITE            2
+#define QCOW2_OP_COMMIT           3
 
 #define QCOW2_FLAG_OPEN_RDONLY         1
 #define QCOW2_FLAG_OPEN_NO_CACHE       2
@@ -120,6 +123,7 @@ struct qcow2_request {
     QLIST_ENTRY(qcow2_request) list;
     int                     aio_inflight;
     QEMUIOVector            qiov;
+    char *                  top;
 #if DEBUGGING != 0
     int                     id;
     struct timeval          reqtv;
@@ -160,6 +164,7 @@ struct qcow2_state {
 static void qcow2_complete(void *, int);
 static inline void do_aio_read(struct qcow2_state *s, struct qcow2_request *req);
 static inline void do_aio_write(struct qcow2_state *s, struct qcow2_request *req, uint64_t offset);
+static inline void do_commit(struct qcow2_state *s, struct qcow2_request *req);
 
 static int
 qcow2_initialize(struct qcow2_state *s, Error **perr)
@@ -211,6 +216,9 @@ static void qcow2_handle_requests(struct qcow2_state *s)
             case QCOW2_OP_WRITE:
                 do_aio_write(s, req, 0);
                 break;
+            case QCOW2_OP_COMMIT:
+                do_commit(s, req);
+                break;
         }
         pthread_mutex_lock(&lock);
     }
@@ -815,6 +823,54 @@ fail:
     td_complete_request(treq, err);
 }
 
+int
+qcow2_commit(td_driver_t *driver, const char *name)
+{
+    struct qcow2_state *s = (struct qcow2_state *)driver->data;
+    struct qcow2_request *req;
+
+    DBG(TLOG_DBG, "Qcow2: commit %s.\n", name);
+
+    req = alloc_qcow2_request(s);
+    if (!req)
+        return -EBUSY;
+
+    req->top   = strdup(name);
+    req->op    = QCOW2_OP_COMMIT;
+
+    pthread_mutex_lock(&lock);
+    QLIST_INSERT_HEAD(&s->inflight, req, list);
+    pthread_mutex_unlock(&lock);
+
+    qemu_bh_schedule(s->bh);
+
+    return 0;
+}
+
+#define COMMIT_JOB_ID "JIDCOMMIT0"
+
+static inline void
+do_commit(struct qcow2_state *s, struct qcow2_request *req)
+{
+    Error *local_err = NULL;
+    char *node_name;
+
+    node_name = blk_bs(s->conf.blk)->node_name;
+
+    DBG(TLOG_DBG, "Qcow2: block commit %s (node-name: '%s').\n", req->top, node_name);
+
+    qmp_block_commit(COMMIT_JOB_ID, node_name, NULL, NULL, NULL, req->top, NULL,
+            false, false, false, 0, false, BLOCKDEV_ON_ERROR_REPORT,
+            NULL, false, false, true, false, &local_err);
+
+    if (local_err) {
+        DPRINTF("qcow2_commit: error: %s\n", error_get_pretty(local_err));
+        error_free(local_err);
+    }
+
+    free(req->top);
+}
+
 void
 qcow2_debug(td_driver_t *driver)
 {
@@ -831,5 +887,6 @@ struct tap_disk tapdisk_qcow = {
 	.td_queue_write     = qcow2_queue_write,
 	.td_get_parent_id   = qcow2_get_parent_id,
 	.td_validate_parent = qcow2_validate_parent,
+	.td_commit          = qcow2_commit,
 	.td_debug           = qcow2_debug,
 };
