From 8eeac9b1cbf2c1060d2131fdf768dbcaea5d3125 Mon Sep 17 00:00:00 2001
From: Mark Syms <mark.syms@citrix.com>
Date: Mon, 6 Jan 2025 10:10:51 +0000
Subject: [PATCH] CP-35551: stop tapback failing to start without blktap2
 device

Signed-off-by: Mark Syms <mark.syms@cloud.com>
---
 tapback/tapback.c | 33 ---------------------------------
 1 file changed, 33 deletions(-)

diff --git a/tapback/tapback.c b/tapback/tapback.c
index fe04d177..60b5712a 100644
--- a/tapback/tapback.c
+++ b/tapback/tapback.c
@@ -472,32 +472,6 @@ tapback_backend_run(backend_t *backend)
     return err;
 }
 
-static int
-get_tapdev_major()
-{
-    FILE *devices;
-    char buf[128], name[128];
-    int major, err = -ENODEV;
-
-    devices = fopen("/proc/devices", "r");
-    if (!devices)
-        return -errno;
-
-    while (fgets(buf, sizeof(buf), devices)){
-        memset(name, 0, sizeof(name));
-        if (sscanf(buf, "%d %127c", &major, name) == 2){
-            if (!strcmp(name, BLKTAP2_DEVNAME"\n")){
-                tapdev_major = major;
-                err = 0;
-                break;
-            }
-        }
-    }
-
-    fclose(devices);
-    return err;
-}
-
 /**
  * Print tapback's usage instructions.
  */
@@ -593,13 +567,6 @@ int main(int argc, char **argv)
 		goto fail;
 	}
 
-    if (get_tapdev_major()) {
-        WARN(NULL, "Unable to find blktap2 device major\n");
-        err = EINVAL;
-        goto fail;
-    }
-    DBG(NULL, "Identified blktap2 device major = %d\n", tapdev_major);
-
     prog = basename(argv[0]);
 
     opt_debug = 0;
