From: James Davis <james.davis1@citrix.com>
CA-220510: Initialsed vhd_file_table pointer before use

Signed-off-by: James Davis <james.davis1@citrix.com>

diff --git a/vhd/vhd-index.c b/vhd/vhd-index.c
index 60a5467..3573710 100644
--- a/vhd/vhd-index.c
+++ b/vhd/vhd-index.c
@@ -918,6 +918,8 @@ vhd_index_summary(vhdi_name_t *name, uint32_t block)
 	vhdi_context_t vhdi;
 	vhdi_file_table_t files;
 
+	memset(&files, 0, sizeof(vhdi_file_table_t));
+
 	err = vhdi_open(&vhdi, name->index, O_RDWR);
 	if (err)
 		return err;
